Add WebSub Spec (#506)
ballerina-platform/module-ballerina-websub: docs/spec/spec.md

# Specification: Ballerina WebSub Library

_Owners_: @shafreenAnfar @chamil321 @ayeshLK    
_Reviewers_: @shafreenAnfar    
_Created_: 2022/01/31  
_Updated_: 2022/01/31  
_Issue_: [#786](https://github.com/ballerina-platform/ballerina-standard-library/issues/786)

# Introduction

This is the specification for WebSub standard library which is used to implement WebSub compliant `subscriber` services
using [Ballerina programming language](https://ballerina.io/), which is an open-source programming language for the 
cloud that makes it easier to use, combine, and create network services.  

# Content
1. [Overview](#1-overview)
2. [Subscriber](#2-subscriber)
    * 2.1. [Listener](#21-listener)
      * 2.1.1. [Configuration](#211-configuration)
      * 2.1.2. [Initialization](#212-initialization)
      * 2.1.3. [Methods](#213-methods)
    * 2.2 [Subscriber Service](#22-subscriber-service)
      * 2.2.1. [Methods](#221-methods)
        * 2.2.1.1. [onSubscriptionValidationDenied](#2211-onsubscriptionvalidationdenied)
        * 2.2.1.2. [onSubscriptionVerification](#2212-onsubscriptionverification)
        * 2.2.1.3. [onUnsubscriptionVerification](#2213-onunsubscriptionverification)

>         * 2.2.1.1. [onEventNotification](#2214-oneventnotification)

Reviewer: ```suggestion
        * 2.2.1.4. [onEventNotification](#2214-oneventnotification)
```

    *http:ListenerConfiguration;
    decimal gracefulShutdownPeriod = 20;
|};
```

For more details on the available configurations please refer [`http:ListenerConfiguration`](https://lib.ballerina.io/ballerina/http/latest/records/ListenerConfiguration).

#### 2.1.2. Initialization

The `websub:Listener` could be initialized by providing either a port with `websub:ListenerConfiguration` or by
providing an `http:Listener`.
```ballerina
# Initiliazes `websub:Listener` instance.
# ```ballerina
# listener websub:Listener websubListenerEp = check new (9090);
# ```
#
# + listenTo - Port number or a `http:Listener` instance
# + config - Custom `websub:ListenerConfiguration` to be provided to underlying HTTP Listener
# + return - The `websub:Listener` or an `websub:Error` if the initialization failed
public isolated function init(int|http:Listener listenTo, *ListenerConfiguration config) returns websub:Error? 
```

#### 2.1.3. Methods

Following APIs should be available in the `websub:Listener` to dynamically attach `websub:SubscriberService` objects to 
it.
```ballerina
# Attaches the provided `websub:SubscriberService` to the `websub:Listener`.

> # ```ballerina

Reviewer: ```suggestion
```

# check websubListenerEp.attach('service, "/subscriber");

> # ```

Reviewer: ```suggestion
```

Reviewer: Let's remove these inner quotes. Due to the out quotes, inner quotes will not function I think. Check the other places as well

- `service-info.csv` file should be added to `resources/ballerina/websub` directory inside the executable JAR as well as the thin JAR.
- If there is an error while generating the service path, then it should result in a compile-time error since this feature is required to generate a callback URL and without it subscriber service could not be used.

#### 2.2.4. Unsubscribing from the `hub`

As per the WebSub Specification a `subscriber` should be able to subscribe to a particular `topic` in a specific `hub`. 
As the counterpart subscriber should be able to unsubscribe from a previously subscribed `topic` in a specific `hub`. 
Even though the specification does not clearly define when this unsubscription should happen, it is obvious that this 
should happen whenever `subscriber` is terminated.  

The key functionalities expected from the unsubscription flow as follows:  
- Developer should manually enable this feature by configuring `unsubscribeOnShutdown` in `websub:SubscriberServiceConfig` annotation.
- Developer should be able to configure unsubscription verification time-out using `gracefulShutdownPeriod` configuration in `websub:ListenerConfiguration`.

> - Unsubscription flow should initiate whenever graceful stop is invoked in `websub:Listener`.

Reviewer: ```suggestion
- Unsubscription flow should be initiated whenever the graceful stop is invoked in `websub:Listener`.
```
